go/store/prolly/tree/json_chunker.go

```diff
@@ -0,0 +1,260 @@
+func SerializeJsonToAddr(ctx context.Context, ns NodeStore, j sql.JSONWrapper) (Node, error) {
+	if indexedJson, ok := j.(IndexedJsonDocument); ok {
+		return indexedJson.m.Root, nil
+	}
+	jsonBytes, err := types.MarshallJson(j)
+	if err != nil {
+		return Node{}, err
+	}
+
+	jsonChunker, err := newEmptyJsonChunker(ctx, ns)
+	if err != nil {
+		return Node{}, err
+	}
+	jsonChunker.appendJsonToBuffer(jsonBytes)
+	jsonChunker.processBuffer(ctx)
+
+	node, err := jsonChunker.Done(ctx)
+	if err != nil {
+		return Node{}, err
+	}
+	return node, nil
+}
+
+// newEmptyJsonChunker creates a new JsonChunker without a corresponding JsonCursor. This is used when writing
+// a new IndexedJsonDocument that not based on an existing IndexedJsonDocument.
+func newEmptyJsonChunker(ctx context.Context, ns NodeStore) (*JsonChunker, error) {
+	newChunkerFn := newChunker[message.AddressMapSerializer]
+	chunker, err := newChunkerFn(ctx, nil, 1, ns, message.NewAddressMapSerializer(ns.Pool()))
+	if err != nil {
+		return nil, err
+	}
+	scanner := ScanJsonFromBeginning(nil)
+	jChunker := JsonChunker{
+		jCur:     nil,
+		jScanner: &scanner,
+		chunker:  chunker,
+		ns:       ns,
+	}
+	return &jChunker, err
+}
+
+// newJsonChunker creates a new JsonChunker based on an existing IndexedJsonDocument.
+// |jCur| is a cursor into the existing document, pointing to the location of the first change.
+// |nextKey| is the location in the document of the next value to be written.
+func newJsonChunker(ctx context.Context, jCur *JsonCursor, nextKey jsonLocation, ns NodeStore) (*JsonChunker, error) {
+	newChunkerFn := newChunker[message.AddressMapSerializer]
+	chunker, err := newChunkerFn(ctx, jCur.cur.parent, 1, ns, message.NewAddressMapSerializer(ns.Pool()))
+	if err != nil {
+		return nil, err
+	}
+
+	// Copy the original bytes so that the JsonChunker's buffer doesn't point into JsonCursor's buffer.
+	initialBytes := bytes.Clone(jCur.jsonScanner.jsonBuffer[:jCur.jsonScanner.previousValueOffset])
+	scanner := JsonScanner{
+		jsonBuffer:   initialBytes,
+		currentPath:  nextKey.Clone(),
+		valueOffset:  len(initialBytes),
+		firstElement: jCur.jsonScanner.firstElement,
+	}
+
+	jChunker := JsonChunker{
+		jCur:     jCur,
+		jScanner: &scanner,
+		chunker:  chunker,
+		ns:       ns,
+	}
+
+	jChunker.writeKey(nextKey)
+
+	return &jChunker, nil
+}
+
+// writeKey adds a new key to end of the JsonChunker's buffer. If required, it also adds a comma before the key.
+// If the path points to an array, no key will be written, but the comma will still be written if required.
+func (j *JsonChunker) writeKey(keyPath jsonLocation) {
+	finalPathElement, isArray := keyPath.getLastPathElement()
+
+	isFirstValue := j.jScanner.firstElementOrEndOfEmptyValue()
+
+	if !isFirstValue {
+		j.appendJsonWithoutSplitting([]byte{','})
+	}
+	if !isArray {
+		j.appendJsonWithoutSplitting([]byte(fmt.Sprintf(`"%s":`, finalPathElement)))
```

Comment:
Definitely somewhat confusing on how this fits together with the existing jScanner state and it's currentPath, and why it's safe specifically in this one case immediately after the chunker has been constructed with this specific keyPath and never safe again.

Reply:
You're right. Given that this function is only called in one place and isn't called anywhere else, I decided to just inline it and add comments.